hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngradeLegacy.java

```diff
@@ -709,10 +710,11 @@ void testNeedsUpgrade() {
     assertTrue(shouldUpgrade);
 
     // assert upgrade for table version 8 from table version 6 with auto upgrade set to false
+    // should return false
     when(writeConfig.autoUpgrade()).thenReturn(false);
     shouldUpgrade = new UpgradeDowngrade(metaClient, writeConfig, context, null)
         .needsUpgrade(HoodieTableVersion.EIGHT);
-    assertTrue(shouldUpgrade);
+    assertFalse(shouldUpgrade);
```

Comment:
Have you checked the existing test cases cover all combinations for `needsUpgrade`?

Reply:
I tried actually adding a recent test for this in the `TestMORDataSource`